crates/accelerate/src/synthesis/linear_phase/cnot_phase_synth.rs

```diff
@@ -0,0 +1,292 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::synthesis::linear::pmh::_synth_cnot_count_full_pmh;
+use ndarray::Array2;
+use numpy::PyReadonlyArray2;
+use pyo3::{prelude::*, types::PyList};
+use qiskit_circuit::circuit_data::CircuitData;
+use qiskit_circuit::operations::{Param, StandardGate};
+use qiskit_circuit::Qubit;
+use smallvec::smallvec;
+use std::f64::consts::PI;
+
+/// This function implements a Gray-code inspired algorithm of synthesizing a circuit
+/// over CNOT and phase-gates with minimal-CNOT for a given phase-polynomial.
+/// The algorithm is described as "Gray-Synth" algorithm in Algorithm-1, page 12
+/// of paper "https://arxiv.org/abs/1712.01859".
+#[pyfunction]
+#[pyo3(signature = (cnots, angles, section_size=2))]
+pub fn synth_cnot_phase_aam(
+    py: Python,
+    cnots: PyReadonlyArray2<u8>,
+    angles: &Bound<PyList>,
+    section_size: Option<i64>,
+) -> PyResult<CircuitData> {
+    let s = cnots.as_array().to_owned();
+    let num_qubits = s.nrows();
+    let mut s_cpy = s.clone();
+    let mut instructions = vec![];
+
+    let mut rust_angles: Vec<String> = angles
+        .iter()
+        .filter_map(|data| data.extract::<String>().ok())
+        .collect();
```

Comment:
I know the Python code does the same, but this seems highly volatile and we should change this: the input type should be properly differentiated in _either_ floats _or_ strings, but not a mix. You can leave this for now @MozammilQ but maybe we can find a better solution @alexanderivrii  and @ShellyGarion  🙂

Reply:
Indeed, it's quite strange that the API of `angles` can accept strings e.g. `s, t, sdg, tdg or z` or `float`.
in the [test](https://github.com/Qiskit/qiskit/blob/main/test/python/synthesis/test_cnot_phase_synthesis.py) we see that the angles are only strings, e.g. `s, t, sdg, tdg or z`. It would be good to add some tests of angles which are `float` to check that the code works well for any angle.

Reply:
@ShellyGarion , 
I had already taken care of this by `angles = [angle if isinstance(angle, str) else f"{angle}" for angle in angles]` in `cnot_phase_synth.py`. This converts all occurrences of float in the list of angles to String  which rust subsequently accepts.

Anyways, I have added a test. Take a note of the change in circuit. I have just changed the circuit in the docstring to the actual circuit I am getting here locally, but not have changed the circuit in the code, just to make sure the code works for the circuit generated from the older python implementation of `synth_cnot_phase_aam`, there by proving the circuit generated by rust is merely "equivalent" to the circuit generated by python version and not different.

Nevertheless, I am surprised because there is no random steps in the rust or the python implementation of the algorithm, so both the rust and python implementation should produce exact same circuit, but this is not the case.
Maybe, the difference is because when I append the output of `synth_pmh` I just do `.rev()` but not the actual `.inverse()` in python space as it has been done originally in python implementation.
I decided to do only `.rev()` thinking inverse of a `cx` is itself. So, just reversing the strings of `cx`s in the circuit should be equivalent to doing `.inverse()` on QuantumCircuit in python space. Please, guide me if you feel I am lost :)